ddht/rpc.py

```diff
@@ -0,0 +1,253 @@
+
+    return response
+
+
+TParams = TypeVar("TParams")
+TResult = TypeVar("TResult")
+
+
+class RPCError(Exception):
+    ...
+
+
+class RPCHandler(RPCHandlerAPI, Generic[TParams, TResult]):
+    async def __call__(self, request: RPCRequest) -> RPCResponse:
+        try:
+            params = self.extract_params(request)
+            result = await self.do_call(params)
+        except RPCError as err:
+            return self.generate_error_response(request, str(err))
+        else:
+            return self.generate_success_response(request, result)
+
+    @abstractmethod
+    async def do_call(self, params: TParams) -> TResult:
+        ...
+
+    def extract_params(self, request: RPCRequest) -> TParams:
+        return request.get("params", [])  # type: ignore
+
+    def generate_success_response(
+        self, request: RPCRequest, result: Any
+    ) -> RPCResponse:
+        return generate_response(request, result, None)
+
+    def generate_error_response(self, request: RPCRequest, error: str) -> RPCResponse:
+        return generate_response(request, None, error)
+
+
+class UnknownMethodHandler(RPCHandlerAPI):
+    async def __call__(self, request: RPCRequest) -> RPCResponse:
+        return generate_response(
+            request, None, f"Unknown RPC method: {request['method']}"
+        )
+
+
+fallback_handler = UnknownMethodHandler()
+
+
+class RPCServer(Service):
+    logger = logging.getLogger("alexandria.rpc.RPCServer")
+    _handlers: Dict[str, RPCHandlerAPI]
+
+    def __init__(
+        self, ipc_path: pathlib.Path, handlers: Dict[str, RPCHandlerAPI]
+    ) -> None:
+        self.ipc_path = ipc_path
+        self._handlers = handlers
+        self._serving = trio.Event()
+
+    async def wait_serving(self) -> None:
+        await self._serving.wait()
+
+    async def run(self) -> None:
+        self.manager.run_daemon_task(self.serve, self.ipc_path)
+        try:
+            await self.manager.wait_finished()
+        finally:
+            self.ipc_path.unlink()
+
+    async def execute_rpc(self, request: RPCRequest) -> str:
+        method = request["method"]
+
+        self.logger.debug("RPCServer handling request: %s", method)
+
+        handler = self._handlers.get(method, fallback_handler)
+        try:
+            response = await handler(request)
+        except Exception as err:
+            self.logger.error("Error handling request: %s  error: %s", request, err)
+            self.logger.debug("Error handling request: %s", request, exc_info=True)
+            response = generate_response(request, None, f"Unexpected Error: {err}")
+        finally:
+            return json.dumps(response)
+
+    async def serve(self, ipc_path: pathlib.Path) -> None:
+        self.logger.info("Starting RPC server over IPC socket: %s", ipc_path)
+
+        with trio.socket.socket(trio.socket.AF_UNIX, trio.socket.SOCK_STREAM) as sock:
+            # TODO: unclear if the following stuff is necessary:
```

Comment:
[The man page](https://linux.die.net/man/2/bind) is a little vague on this, but I'm pretty sure `SO_REUSEADDR` has no effect on unix domain sockets. [net/unix/af_unix.c](https://github.com/torvalds/linux/blob/master/net/unix/af_unix.c) never touches `sk_reuse` before deciding to send `EADDRINUSE`.